### ISSUE TITLE
Ability to catch exceptions

### ISSUE THREAD

**Issue by chriseth** [ethereum/solidity#1505]
A general `try { ... } catch`-statement should be available to catch exceptions.

How to implement it: The "invalid label" has to become dynamic and somehow either the stack height has to be stored or it will be impossible to access local variables. Furthermore, we have to decide whether local effects should be reverted or not.

**Comment by federicobond** [comment]
Would it be possible to attach a label to the catch statement? Too general catch clauses have proven to be a constant source of bugs in many programs.

Perhaps these tags could be translated to simple `errno`-like integers to be stored in a special location and the `catch label {}` could check against this number before running the catch clause, or continue the stack unwinding otherwise.

Sure, you don't need language support to do this, but given how difficult it is to get this kind of error handling right, I think it would be a welcome addition.

**Comment by chriseth** [comment]
The main problem to overcome is how to handle cross-call exceptions.

**Comment by pirapira** [comment]
@chriseth This can be solved by passing two return pointers on the stack.  One for normal returning, and the other for exception throwing.

**Comment by axic** [comment]
With `REVERT` once the `CALL` opcode is able to return 3 states (success, exception, revert) we could consider introducing exception catching for methods:
```
try {
  Wallet(addr1).transferTo(addr2);
}
catch (e)
{
  if (e == "nofunds")
    // yay, we ran out of monez
}
```

This would be backwards compatible, because the higher level method still throws an exception, but the try clause will execute if the `CALL` failed with `REVERT`.

**Comment by chriseth** [comment]
@axic Do you think cross-internal-call exception catching can be done in ewasm?

**Comment by fulldecent** [comment]
I am a fan of the MVP:

```solidity
try {
  Wallet(addr1).transferTo(addr2);
} catch {
  // something didn't work
}
```

---

In other news, I am not a fan of storing English-language error messages inside the blockchain. So I think this MVP is a final product.

**Comment by SilentCicero** [comment]
Hey all,

I'm trying to impliment this for the Gnosis MultiSend, such that if one of the transactions fails, the MultiSend skips and continues with the next transaction, thoughts?

I suppose I just remove the `revert` line, and that should do it right?

```
pragma solidity ^0.5.0;


/// @title Multi Send - Allows to batch multiple transactions into one.
/// @author Nick Dodson - <nick.dodson@consensys.net>
/// @author Gonçalo Sá - <goncalo.sa@consensys.net>
/// @author Stefan George - <stefan@gnosis.pm>
contract MultiSend {

    /// @dev Sends multiple transactions and reverts all if one fails.
    /// @param transactions Encoded transactions. Each transaction is encoded as a 
    ///                     tuple(operation,address,uint256,bytes), where operation 
    ///                     can be 0 for a call or 1 for a delegatecall. The bytes 
    ///                     of all encoded transactions are concatenated to form the input.
    function multiSend(bytes memory transactions)
        public
    {
        // solium-disable-next-line security/no-inline-assembly
        assembly {
            let length := mload(transactions)
            let i := 0x20
            for { } lt(i, length) { } {
                let operation := mload(add(transactions, i))
                let to := mload(add(transactions, add(i, 0x20)))
                let value := mload(add(transactions, add(i, 0x40)))
                let dataLength := mload(add(transactions, add(i, 0x80)))
                let data := add(transactions, add(i, 0xa0))
                let success := 0
                switch operation 
                case 0 { success := call(gas, to, value, data, dataLength, 0, 0) }
                case 1 { success := delegatecall(gas, to, data, dataLength, 0, 0) }
                if eq(success, 0) { revert(0, 0) }
                i := add(i, add(0xa0, mul(div(add(dataLength, 0x1f), 0x20), 0x20)))
            }
        }
    }
}
```

**Comment by chriseth** [comment]
Has been implemented for external calls in 0.6.0.
